subprojects/model-core/src/main/java/org/gradle/internal/properties/annotations/TypeMetadataWalker.java

```diff
@@ -21,20 +21,80 @@
 import java.lang.annotation.Annotation;
 import java.util.function.Supplier;
 
+/***
+ * A generalized Type Metadata Walker for Gradle types or instances.
+ *
+ * A walker visits all properties of a type or an instance.
+ */
 public interface TypeMetadataWalker<T> {
+
+    /**
+     * A factory method for a walker that can visit an instance.
+     * A walker is lightweight object so there is no need to cache it, but it holds no state, so it can be reused.
```

Comment:
This is not strictly true, as the `TypeMetadataStore` does hold state.

Whatever we write here, I think this info should go to the Javadoc of the interface itself. That way it doesn't have to be repeated twice.